<div class="alert alert-block alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h5> Sprint 7 Project:  Intro to Machine Learning </h5>
    <p>(A) Import the data files and model libraries
    <p>=>>  (1) import library for TRAIN_TEST_SPLIT
    <p>=>>  (2) import libraries for model training (DecisionTreeClassifier, RandomForestClassifier, LogisticRegression)
    <p>=>>  (3) import libraries for model accuracy/quality
    <p>  <br>    
    <p>(B) Study the data 
    <p>=>>  (1) import '/datasets/users_behavior.csv' (Megaline Subscriber behavior data)
    <p>=>>  (2) df.info()
    <p>=>>  (3) df.isna()
    <p>=>>  (4) df.isinf()
    <p>=>>  (5) Fill all NAN/INF values, if any are found
    <p>  <br>        
    <p>(C) Split the data, check the shape(), setup and tune hyperparameters, fit and train the model
    <p>=>>  (1) Split the data into a TRAINING, VALIDATION and TEST set (3:1:1 Model - 60%, 20% 20%)
    <p>=>>  (2) DF.SHAPE() check to verify that data split correctly into a TRAINING, VALIDATION and TEST set (3:1:1 Model - 60%, 20% 20%)
    <p>=>>  (3) Create the FEATURES and TARGET variables from TRAIN, VALID and TEST datasets
    <p>=>>  (4) SHAPE() validation of the FEATURES and TARGET variables    
    <p>=>>  (5) Fit and train the model(s) 
    <p>         >>>>>>>    (a)  DecisionTreeClassifier
    <p>         >>>>>>>    (b)  RandomForestClassifier
    <p>         >>>>>>>    (c)  LogisticRegression
    <p>=>>  (6) Setup and tune parameters (max_depth, random_state, n_estimators)
    <p>  <br>        
    <p>(D) Assess model quality
    <p>=>>  (1) Check model accuracy/quality (accuracy_score, model.score) for each model
    <p>=>>  (2) Choose the BEST model 
    <p>  <br>            
 </div>

In [1]:
## Sprint 7 - (A)

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier     
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
## Sprint 7 - (B-1, B-2)

df = pd.read_csv('/datasets/users_behavior.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [3]:
df.head(10)

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0

In [4]:
## Sprint 7 - (B-3)

df.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

In [5]:
## Sprint 7 - (B-4)

df_inf = np.isinf(df)
df_inf_counts = df_inf.sum(axis=0)
print(df_inf_counts)

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64


In [6]:
## Sprint 7 - (B-5)

## No need to fill, no NULL or INF values found (see B-3, B-4)

In [7]:
## Sprint 7 - (C-1) 
##  Split the data into a TRAINING, VALIDATION and TEST set (3:1:1 Model - 60%, 20% 20%)

##  First split:  divide data into 60% train and 40% temp (this will become validation and test sets)
df_train, df_temp = train_test_split(df, test_size=0.4, random_state=77)

##  Second split:  divide df_temp into df_valid and df_test into 20% validation and 20% test sets
df_valid, df_test = train_test_split(df_temp, test_size=0.5, random_state=77)

In [8]:
## Sprint 7 - (C-2) 
##   DF.SHAPE() check to verify that data split correctly into TRAIN, VALID and TEST set (3:1:1 Model - 60%, 20% 20%)

print('df_train shape:', df_train.shape)
print()
print('df_temp shape:', df_temp.shape)
print()
print('df_valid shape:', df_valid.shape)
print()
print('df_test shape:', df_test.shape)

df_train shape: (1928, 5)

df_temp shape: (1286, 5)

df_valid shape: (643, 5)

df_test shape: (643, 5)


In [9]:
## Sprint 7 - (C-3, C-4) 
##   Create the FEATURES and TARGET variables from TRAIN, VALID and TEST datasets
##     SHAPE() validation of the FEATURES and TARGET variables

df_train_features = df_train.drop(['is_ultra'], axis=1)
df_train_target = df_train['is_ultra']

df_valid_features = df_valid.drop(['is_ultra'], axis=1)
df_valid_target = df_valid['is_ultra']

df_test_features = df_test.drop(['is_ultra'], axis=1)
df_test_target = df_test['is_ultra']

print('df_train_features:', df_train_features.shape)
print('df_train_target:', df_train_target.shape)
print()
print('df_valid_features:', df_valid_features.shape)
print('df_valid_target:', df_valid_target.shape)
print()
print('df_test_features:', df_test_features.shape)
print('df_test_target:', df_test_target.shape)

df_train_features: (1928, 4)
df_train_target: (1928,)

df_valid_features: (643, 4)
df_valid_target: (643,)

df_test_features: (643, 4)
df_test_target: (643,)


In [10]:
df_train_features.head(10)

calls  minutes  messages   mb_used
2701   43.0   173.90      60.0  23233.82
794     6.0    46.40       0.0    814.35
3151   84.0   622.60      32.0  16318.74
1381   63.0   403.32      65.0  16784.02
2109   67.0   431.79       5.0   6287.89
714    60.0   376.33       1.0  13479.08
2485   26.0   167.75      22.0   8673.33
765    45.0   320.44     117.0  35689.60
2404   62.0   413.42     152.0  16139.12
2729   31.0   222.30      65.0  16914.19

In [11]:
## Sprint 7 - (C-5, C-6, D-1) 
##   Fit and train the model(s)
##      (a) DecisionTreeClassifier

##  Create, Train and Predict a model
##  Create a loop for max_depth from 1 to 9
for depth in range(1, 9):
    model = DecisionTreeClassifier(random_state=77, max_depth=depth)
    
    ##  Train the model
    model.fit(df_train_features, df_train_target)
    
    ##  Find Predictions using the VALID dataset
    predictions_valid = model.predict(df_valid_features)
    
    print("max_depth =", depth, ": ", end="")
    print(accuracy_score(df_valid_target, predictions_valid))

max_depth = 1 : 0.7387247278382582
max_depth = 2 : 0.776049766718507
max_depth = 3 : 0.7838258164852255
max_depth = 4 : 0.7791601866251944
max_depth = 5 : 0.7884914463452566
max_depth = 6 : 0.7869362363919129
max_depth = 7 : 0.7916018662519441
max_depth = 8 : 0.7822706065318819


In [12]:
## Sprint 7 - (C-5, C-6, D-1)
##   Fit and train the model(s)
##      (b) RandomForestClassifier

##  Create a hyperparameter range from 1 to 30
##  Create, Train and Predict a model
best_score = 0
best_est = 0
for est in range(1, 30):
    ##  set number of trees
    model = RandomForestClassifier(random_state=77, n_estimators=est)
    
    ##  Train on TRAIN model
    model.fit(df_train_features, df_train_target)
    
    ##  Create an accuracy score using the VALID set
    score = model.score(df_valid_features, df_valid_target)
    
    ##  Find BEST score. Display number of estimators it takes to get the BEST score.
    if score > best_score:
        best_score = score
        best_est = est
        
##  Print the BEST score model
print("Using Range 1 - 30")
print("Accuracy of the BEST model, using VALID set (n_estimators = {}):  {}".format(best_est, best_score))


Using Range 1 - 30
Accuracy of the BEST model, using VALID set (n_estimators = 14):  0.7931570762052877


In [13]:
## Sprint 7 - (C-5, C-6, D-1)
##   Fit and train the model(s)
##      (b) RandomForestClassifier

##  Create a hyperparameter range from 1 to 50
##  Create, Train and Predict a model
best_score = 0
best_est = 0
for est in range(1, 50):
    ##  set number of trees
    model = RandomForestClassifier(random_state=77, n_estimators=est)
    
    ##  Train on TRAIN model
    model.fit(df_train_features, df_train_target)
    
    ##  Create an accuracy score using the VALID set
    score = model.score(df_valid_features, df_valid_target)
    
    ##  Find BEST score. Display number of estimators it takes to get the BEST score.
    if score > best_score:
        best_score = score
        best_est = est
        
##  Print the BEST score model
print("Using Range 1 - 50")
print("Accuracy of the BEST model, using VALID set (n_estimators = {}):  {}".format(best_est, best_score))


Using Range 1 - 50
Accuracy of the BEST model, using VALID set (n_estimators = 48):  0.8009331259720062


In [14]:
## Sprint 7 - (C-5, C-6, D-1)
##   Fit and train the model(s)
##      (b) RandomForestClassifier

##  Create a hyperparameter range from 1 to 120
##  Create, Train and Predict a model
best_score = 0
best_est = 0
for est in range(1, 120):
    ##  set number of trees
    model = RandomForestClassifier(random_state=77, n_estimators=est)
    
    ##  Train on TRAIN model
    model.fit(df_train_features, df_train_target)
    
    ##  Create an accuracy score using the VALID set
    score = model.score(df_valid_features, df_valid_target)
    
    ##  Find BEST score. Display number of estimators it takes to get the BEST score.
    if score > best_score:
        best_score = score
        best_est = est
        
##  Print the BEST score model
print("Using Range 1 - 120")
print("Accuracy of the BEST model, using VALID set (n_estimators = {}):  {}".format(best_est, best_score))


Using Range 1 - 120
Accuracy of the BEST model, using VALID set (n_estimators = 92):  0.80248833592535


In [15]:
## Sprint 7 - (C-5, C-6, D-1)
##   Fit and train the model(s)
##      (c) LogisticRegression

## Initialize the model
model = LogisticRegression(random_state=77, solver='liblinear')

## Train model using the TRAIN set
model.fit(df_train_features, df_train_target)

## Calculate ACCURACY SCORE using the TRAIN set
score_train = model.score(df_train_features, df_train_target)

## Calculate ACCURACY SCORE using the VALID set
score_valid = model.score(df_valid_features, df_valid_target)

##  Print SCORES
print("LogisticRegression SCORE for TRAIN set:", score_train)
print("LogisticRegression SCORE for VALID set:", score_valid)


LogisticRegression SCORE for TRAIN set: 0.6996887966804979
LogisticRegression SCORE for VALID set: 0.7060653188180405


In [20]:
## Sprint 7 - (C-5, D-2)
##   Fit and train the FINAL model
##      (b) RandomForestClassifier

## Use the BEST ESTIMATOR from previous tested models
final_model = RandomForestClassifier(random_state=77, n_estimators=best_est)
    
##  Train on both the TRAIN and VALID data sets
final_model.fit(pd.concat([df_train_features, df_valid_features]), pd.concat([df_train_target, df_valid_target]))

##  Find Predictions using the DF_TEST_FEATURES dataset
predictions_test = final_model.predict(df_test_features)
    
accuracy_test = accuracy_score(df_test_target, predictions_test)

print("Accuracy Score of the FINAL model, using TEST set (best estimator = {}):  {}".format(best_est, accuracy_test))


Accuracy Score of the FINAL model, using TEST set (best estimator = 92):  0.8273716951788491


<div class="alert alert-block alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h5> Project Summary: </h5>
    <p>(1) A model with the highest possible accuracy has been created, with an accuracy score above the 0.75 threshold
    <p>(2) The FINAL model used RandomForestClassifier to train and predict using the DF_TEST_FEATURES dataset
    <p>(3) Following are some of the model summary notes:
    <p>  <br>    
    <p>=>  (a) Three different classification models were tested to determine the BEST estimator model to use
    <p>=>  (b) DecisionTreeClassifier, RandomForestClassifier and LogisticRegression
    <p>=>  (c) The best accuracy score (0.7916018662519441) from the DecisionTreeClassifier was yielded by a max depth of 7. Other depths higher than 7 did not yield better results.
    <p>=>  (d) The RandomForestClassifier had the most flexible hyperparameter tuning, ranging from 30, 50 and 120 estimators.
    <p>=>  (e) The larger the number of estimators tested, the better the accuracy score evaluated. Larger estimator numbers also meant longer evaluation times before the score was calculated.
    <p>=>  (f) This classifier produced the BEST training set accuracy score (0.80248833592535) from 92 estimators in a range of 1 to 120.
    <p>=>  (g) The LogisticRegression model was the least accurate for this dataset, with a score of 0.6996887966804979 for the TRAIN dataset, and 0.7060653188180405 for the VALIDATION dataset.
    <p>(4) Based on all the conducted model tuning and testing, the FINAL model selected was the RandomForestClassifier
    <p>=>  (a) This FINAL model accuracy passed the threshold of 0.75 when tested using the TEST dataset        
    <p>=>  (b) The FINAL model was fit to combined datasets of the original TRAIN and VALID datasets        
    <p>=>  (c) The FINAL accuracy score yielded 0.8273716951788491 from this model with 92 estimators.
</div>